pkg/operator/controller/status.go

```diff
@@ -40,71 +47,165 @@ func (r *reconciler) syncOperatorStatus() error {
+func (r *reconciler) computeOperatorStatusVersions(oldVersions []configv1.OperandVersion, conditions []configv1.ClusterOperatorStatusCondition) ([]configv1.OperandVersion, string) {
+	oldReleaseVersion := UnknownVersionValue
+	oldIngressControllerImage := UnknownVersionValue
+	for _, opv := range oldVersions {
+		switch opv.Name {
+		case OperatorVersionName:
+			oldReleaseVersion = opv.Version
+		case IngressControllerVersionName:
+			oldIngressControllerImage = opv.Version
+		}
+	}
+
+	operatorAvailable := false
+	operatorProgressing := false
+	for _, cond := range conditions {
+		switch cond.Type {
+		case configv1.OperatorAvailable:
+			if cond.Status == configv1.ConditionTrue {
+				operatorAvailable = true
+			}
+		case configv1.OperatorProgressing:
+			if cond.Status == configv1.ConditionTrue {
+				operatorProgressing = true
+			}
 		}
+	}
 
-		return nil, nil, fmt.Errorf(
-			"error getting Namespace %s: %v", ns.Name, err)
+	releaseVersion := r.OperatorReleaseVersion
+	ingressControllerImage := r.IngressControllerImage
+	// We need to report old version until the operator fully transitions to the new version.
+	// https://github.com/openshift/cluster-version-operator/blob/master/docs/dev/clusteroperator.md#version-reporting-during-an-upgrade
+	if !operatorAvailable || operatorProgressing {
+		releaseVersion = oldReleaseVersion
+		ingressControllerImage = oldIngressControllerImage
 	}
 
-	ingressList := &operatorv1.IngressControllerList{}
-	if err := r.client.List(context.TODO(), ingressList, client.InNamespace(r.Namespace)); err != nil {
-		return nil, nil, fmt.Errorf("failed to list IngressControllers: %v", err)
+	// Generate version change messages
+	versionChangeMessages := []string{}
+	if releaseVersion != r.OperatorReleaseVersion {
+		versionChangeMessages = append(versionChangeMessages, fmt.Sprintf("Moving to release version %q.", r.OperatorReleaseVersion))
+	}
+	if ingressControllerImage != r.IngressControllerImage {
+		versionChangeMessages = append(versionChangeMessages, fmt.Sprintf("Moving to ingress-controller image version %q.", r.IngressControllerImage))
 	}
 
-	return ns, ingressList.Items, nil
+	return []configv1.OperandVersion{
+		{
+			Name:    OperatorVersionName,
+			Version: releaseVersion,
+		},
+		{
+			Name:    IngressControllerVersionName,
+			Version: ingressControllerImage,
+		},
+	}, strings.Join(versionChangeMessages, "\n")
+}
+
+// updateOperatorStatusProgressingCondition updates operator's progressing condition with the version change message.
+func updateOperatorStatusProgressingCondition(message string, conditions []configv1.ClusterOperatorStatusCondition) {
+	for i := range conditions {
+		if conditions[i].Type == configv1.OperatorProgressing {
+			conditions[i].Status = configv1.ConditionTrue
+			conditions[i].Reason = "Reconciling"
+			if len(conditions[i].Message) != 0 {
+				conditions[i].Message += "\n"
+			}
+			conditions[i].Message += message
```

Comment:
It would be better to use `setOperatorStatusCondition` to make sure the last transition time is updated.